### PR TITLE
feat: add HEXPIRE and FIELDEXPIRE

### PR DESCRIPTION
**Pull request by NegatioN**
Continuation of: https://github.com/dragonflydb/dragonfly/pull/3780
To implement issue: https://github.com/dragonflydb/dragonfly/issues/3027 
Here's my suggestion for how to add the `FIELDEXPIRE` and `HEXPIRE` commands. I tried to split it up similarly to how `FieldTtl` does generic operations across `set_family` and `hset_family`.

I've not added support for the optional prefixes `[NX | XX | GT | LT]`, as I see few other commands do it, and I also don't really need it. (Can probably be further work?)

Responses from both commands are (per element queried): `-2` for non-existing key/field, `0` if ttl is 0, and ttl gets set, normal success=`1`

I have a few questions at this point:
1. I might need help with the flags that are set for the given commands. I frankly just set something that seems to work, and compile.
2. I'm also unsure about the `acl`-settings for the commands.

Open for comments. Tagging @romange @kostasrim, since you're already familiar with the previous PR.

### CODE REVIEW

File: src/server/hset_family.cc [hunk 1]
```diff
@@ -725,6 +725,23 @@ void HGetGeneric(CmdArgList args, ConnectionContext* cntx, uint8_t getall_mask)
...
+  auto& db_slice = op_args.GetDbSlice();
+  auto op_res = db_slice.FindMutable(op_args.db_cntx, key, OBJ_HASH);
+
+  if (!op_res) {
+    if (op_res.status() == OpStatus::KEY_NOTFOUND) {
+      std::vector<long> res(values.size(), -2);
+      return res;
+    }
+    return op_res.status();
+  }
```
Comment (kostasrim) [inline src/server/hset_family.cc:739]: `If (!op_res) return op_res.status()`. The reason is because if `KEY_NOT_FOUND` we need to return an error and not an array. Also if key has `WRONG TYPE` we again need to return an error and not an array. So returning `op_res.status()` satisfies that

**Review by kostasrim — COMMENTED** [review]
Hi @NegatioN and thaaaank you for your contribution!

I left some comments mostly about adding a couple of test cases, some corner cases and some nits to avoid code duplication :) 

Thank you again :rocket: :100:

File: src/server/hset_family.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (NegatioN) [reply src/server/hset_family.cc:739]: Redis seems to return an array of `-2` if the key doesn't exist, shouldn't this command strive for parity on that front?

File: src/server/hset_family.cc [hunk 2]
```diff
@@ -1276,4 +1339,41 @@ int32_t HSetFamily::FieldExpireTime(const DbContext& db_context, const PrimeValu
...
+  for (size_t i = 0; i < values.size(); i++) {
+    string_view field = ToSV(values[i]);
+    auto it = sm->Find(field);
+    if (it != sm->end()) {
+      it.SetExpiryTime(ttl_sec);
+      res.emplace_back(ttl_sec == 0 ? 0 : 1);
+    } else {
+      res.emplace_back(-2);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (NegatioN) [reply src/server/hset_family.cc:1373]: So, is writing higher order functions to solve that alright? It's identical, except the part where you have `StringMap* sm` and `StringSet* ss`.
Without higher order functions or operating on the `DenseSet` directly somehow, I'm not sure how I can unify that entire block, only parts of it.

File: src/server/hset_family.cc [hunk 3]
```diff
@@ -808,6 +825,50 @@ void HSetFamily::HExists(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+void HSetFamily::HExpire(CmdArgList args, ConnectionContext* cntx) {
+  CmdArgParser parser{args};
+  string_view key = parser.Next();
+  string_view ttl_str = parser.Next();
+  uint32_t ttl_sec;
+  constexpr uint32_t kMaxTtl = (1UL << 26);
+  if (!absl::SimpleAtoi(ttl_str, &ttl_sec) || ttl_sec == 0 || ttl_sec > kMaxTtl) {
```
[the remark this answers is not in the document]
Comment (NegatioN) [reply src/server/hset_family.cc:834]: Hm... They're similar, but not identical?
HExpire syntax is: `HEXPIRE key sec FIELDS 1 field1` while my understanding was that FieldExpire syntax was supposed to be `FIELDEXPIRE key sec field1`

So, it's possible to unify some of that, but I landed on just partly duplicating it, since it's actually not the same.

[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:834]: > FIELDEXPIRE key sec field1 

I thought it's identical to `HEXPIRE`. IMO no matter how many times I search it on google, I can't seem to find it so if you have a link on that command I would be extremely happy.

As for the rest, I would just add a bool at the end of the common function `with_fields` and parse the keyword `FIELDS` on demand based on what the caller of the function decides (hexpire/fieldexpire)

File: src/server/hset_family.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:1373]: it's same boilerplate on different data type. You can make it a template :)

File: src/server/hset_family.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:739]: I thought I replied here. I am surprised cause redis always returns key not found for these cases. I tried it myself and you are right. What's even funnier is that they return `-2` per `field` :face_exhaling: :facepalm:

File: src/server/hset_family.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (NegatioN) [reply src/server/hset_family.cc:834]: The [link to HExpire](https://redis.io/docs/latest/commands/hexpire/).
But you mean documentation for `FIELDEXPIRE`? I don't think that exists anywhere yet, and is just what @romange sketched out in https://github.com/dragonflydb/dragonfly/issues/3027

[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:834]: no no I know what `hexpire` is, I could not find `fieldexpire` but now it makes sense. Then I would model it after `hexpire` and add the prerequisite of `FIELDS`. That way the parsing will be identical

[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:834]: @NegatioN Hi! Could I sincerely ask you to `remove the FIELDS delimeter` from `FIELDEXPIRE` ? I synced internally and we decided to diverge them. Furthermore, on the implementation side, you can just add an extra `bool` that conditionally parses the `fields` member

[the remark this answers is not in the document]
Comment (NegatioN) [reply src/server/hset_family.cc:834]: I'll try to join the parsing into one function with a bool to optionally include checking for `FIELDS N`.

Just to clarify: What do you mean by "Remove the FIELDS delimiter" from `FIELDEXPIRE` and `we decided to diverge them`?
To my eyes, `FIELDS N` were never a part of `FIELDEXPIRE`, so there's nothing to remove? But I might be misunderstanding you.

[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:834]: My proposal was to make the two commands have a similar format but after discussing this internally we decided to keep them as is. That is, `FIELDS N` won't be part of `FIELDEXPIRE` and only part of `HEXPIRE`. So my end proposal is that since the parsing is only different for `FIELDS N` we can add a boolean to the parsing function and parse that field optionally. That way we have a common function that can be used by both of the commands

[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:834]: > so there's nothing to remove? 

yes nothing to remove. Only merge the two parsing functions and parse the `FIELDS N` optinally by a bool argument

File: src/server/generic_family.cc [hunk 4]
```diff
@@ -1261,6 +1281,33 @@ void GenericFamily::Restore(CmdArgList args, ConnectionContext* cntx) {
...
 
+void GenericFamily::FieldExpire(CmdArgList args, ConnectionContext* cntx) {
+  CmdArgParser parser{args};
+  string_view key = parser.Next();
+  string_view ttl_str = parser.Next();
+  uint32_t ttl_sec;
+  if (!absl::SimpleAtoi(ttl_str, &ttl_sec) || ttl_sec == 0 || ttl_sec > kMaxTtl) {
+    return cntx->SendError(kInvalidIntErr);
+  }
+  CmdArgList fields = parser.Tail();
```
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/generic_family.cc:1292]: sure no issue with that

File: src/server/hset_family_test.cc [hunk 5]
```diff
@@ -383,6 +389,33 @@ TEST_F(HSetFamilyTest, Issue2102) {
...
+  EXPECT_THAT(Run({"HEXPIRE", "key", "10", "FIELDS", "3", "k0", "k1", "k2"}),
+              RespArray(ElementsAre(IntArg(1), IntArg(1), IntArg(1))));
+  AdvanceTime(10'000);
+  EXPECT_THAT(Run({"HGETALL", "key"}), RespArray(ElementsAre()));
+
+  EXPECT_EQ(CheckedInt({"HSETEX", "key2", "60", "k0", "v0", "k1", "v2"}), 2);
+  EXPECT_THAT(Run({"HEXPIRE", "key2", "10", "FIELDS", "2", "k0", "k1"}),
+              RespArray(ElementsAre(IntArg(1), IntArg(1))));
+  AdvanceTime(10'000);
+  EXPECT_THAT(Run({"HGETALL", "key2"}), RespArray(ElementsAre()));
```
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family_test.cc:403]: one more with a key that does not exist and we should have covered all the cases :)

File: src/server/hset_family.cc [hunk 6]
```diff
@@ -1276,4 +1339,41 @@ int32_t HSetFamily::FieldExpireTime(const DbContext& db_context, const PrimeValu
   }
 }
 
+vector<long> HSetFamily::SetFieldsExpireTime(const OpArgs& op_args, PrimeValue& pv,
+                                             uint32_t ttl_sec, string_view key, CmdArgList values) {
+  DCHECK_EQ(OBJ_HASH, pv.ObjType());
+  vector<long> res;
+  res.reserve(values.size());
+
+  op_args.shard->search_indices()->RemoveDoc(key, op_args.db_cntx, pv);
```
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/hset_family.cc:1347]: I was wrong. We need this because `InitRobj` might change the `inner` pointer which is then used within the `search_indices`. We could conditionally remove only when we create a new `sm` but that's a nit and it won't affect anything :)

**Comment by kostasrim** [comment]
@NegatioN let me know when you addressed all the comments and I will rereview :)

File: src/core/compact_object.h [hunk 7]
```diff
@@ -15,6 +15,7 @@
 #include "core/mi_memory_resource.h"
 #include "core/small_string.h"
 #include "core/string_or_view.h"
+#include "facade/facade_types.h"
```
Comment (romange) [inline src/core/compact_object.h:18]: core should not depend on facade

File: src/core/compact_object.h [hunk 8]
```diff
@@ -150,6 +151,27 @@ class CompactObj {
   using PrefixArray = std::vector<std::string_view>;
   using MemoryResource = detail::RobjWrapper::MemoryResource;
 
+  template <typename O>
```
Comment (romange) [inline src/core/compact_object.h:154]: this is not the right place to add this function

Comment (romange) [reply src/core/compact_object.h:154]: where do you use it?

File: src/server/set_family.h [hunk 9]
```diff
@@ -33,6 +33,9 @@ class SetFamily {
   static int32_t FieldExpireTime(const DbContext& db_context, const PrimeValue& pv,
                                  std::string_view field);
 
+  static std::vector<long> SetFieldsExpireTime(const OpArgs& op_args, PrimeValue& pv,
```
Comment (romange) [inline src/server/set_family.h:36]: our style guide requires that we pass output variables (pv) last and by pointer.

File: src/server/hset_family.h [hunk 10]
```diff
@@ -29,9 +29,14 @@ class HSetFamily {
   static int32_t FieldExpireTime(const DbContext& db_context, const PrimeValue& pv,
                                  std::string_view field);
 
+  static std::vector<long> SetFieldsExpireTime(const OpArgs& op_args, PrimeValue& pv,
```
Comment (romange) [inline src/server/hset_family.h:32]: our style guide requires that we pass output variables (pv) last and by pointer.

File: src/core/compact_object.h (hunk 8 above)
Comment (NegatioN) [reply src/core/compact_object.h:154]: I was actually struggling with the placement of this, so thanks for the feedback.
This would be used in `hset_family` and `set_family`. I didn't see a clear way to facilitate the code-reuse that was suggested other than `compact_object`.

What would be a better fit?

Comment (romange) [reply src/core/compact_object.h:154]: @dranikpg  just introduced family_utils.h - seems like a good place.

File: src/server/hset_family.h (hunk 10 above)
Comment (NegatioN) [reply src/server/hset_family.h:32]: Just to clarify, the signature should be like this?
```cpp
static std::vector<long> SetFieldsExpireTime(const OpArgs& op_args, uint32_t ttl_sec, std::string_view key,
                                               CmdArgList values, PrimeValue* pv)
```

**Comment by NegatioN** [comment]
@kostasrim I think this is ready for a review again now.

File: src/server/family_utils.h [hunk 11]
```diff
@@ -9,11 +9,34 @@
...
 extern "C" {
 #include "redis/sds.h"
 }
 namespace dfly {
 
+template <typename O>
+static std::vector<long> ExpireElements(void* ptr, const facade::CmdArgList values,
+                                        uint32_t ttl_sec) {
+  O* owner = (O*)ptr;
+  std::vector<long> res;
```
Comment (kostasrim) [inline src/server/family_utils.h:23]:
```suggestion
template <typename DenseSet>
static std::vector<long> ExpireElements(DenseSet* owner, const facade::CmdArgList values,
                                        uint32_t ttl_sec) {
  std::vector<long> res;
```

Comment (kostasrim) [reply src/server/family_utils.h:23]: that way you don't need:

1. To cast a `void*` to the respective type
2. You can now call `ExpireElements(arg1, arg2)` without the `<>` syntax. (that is, let the compiler automatically deduce this)

File: src/server/hset_family.h (hunk 10 above)
Comment (kostasrim) [reply src/server/hset_family.h]: I think he wants:

```
static std::vector<long> SetFieldsExpireTime(const OpArgs& op_args, uint32_t ttl_sec, std::string_view key,
                                               CmdArgList values, PrimeValue* pv)
```

Notice last argument, it's not a `reference` but a `pointer`

File: src/server/generic_family.cc [hunk 12]
```diff
@@ -1261,6 +1281,33 @@ void GenericFamily::Restore(CmdArgList args, ConnectionContext* cntx) {
...
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    return OpFieldExpire(t->GetOpArgs(shard), key, ttl_sec, fields);
+  };
+
+  OpResult<vector<long>> result = cntx->transaction->ScheduleSingleHopT(std::move(cb));
+  auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
+  if (result) {
```
Comment (kostasrim) [inline src/server/generic_family.cc:1300]: This could be a function but its fine :)

File: src/server/hset_family.h (hunk 10 above)
Comment (kostasrim) [reply src/server/hset_family.h]: Also @NegatioN I am fine with the PR once this is addressed. The rest of my comments are NIT. 

LGTM! :rocket:

**Comment by NegatioN** [comment]
@kostasrim Done

**Comment by kostasrim** [comment]
LGTM and thank you for the contribution @NegatioN

**Comment by NegatioN** [comment]
btw I have no access to merge the PR, so someone else has to do that for me @kostasrim :)
I want to say thanks for the good feedback from both of you as well, and for staying patient with me even though I have near zero c++ experience.

**Comment by romange** [comment]
I must say it was a pretty good PR for `Zero C++ experience` 😮 😮 😅
